gateway-service-impl/src/main/java/org/hypertrace/gateway/service/logevent/LogEventService.java

```diff
@@ -0,0 +1,155 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadata;
+import org.hypertrace.core.query.service.api.ColumnMetadata;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.ResultSetChunk;
+import org.hypertrace.core.query.service.api.Row;
+import org.hypertrace.core.query.service.client.QueryServiceClient;
+import org.hypertrace.core.serviceframework.metrics.PlatformMetricsRegistry;
+import org.hypertrace.gateway.service.common.AttributeMetadataProvider;
+import org.hypertrace.gateway.service.common.RequestContext;
+import org.hypertrace.gateway.service.common.converters.QueryAndGatewayDtoConverter;
+import org.hypertrace.gateway.service.v1.common.OrderByExpression;
+import org.hypertrace.gateway.service.v1.log.events.LogEvent;
+import org.hypertrace.gateway.service.v1.log.events.LogEventRequest;
+import org.hypertrace.gateway.service.v1.log.events.LogEventResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LogEventService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogEventService.class);
+  private final QueryServiceClient queryServiceClient;
+  private final int requestTimeout;
+  private final AttributeMetadataProvider attributeMetadataProvider;
+
+  private Timer queryExecutionTimer;
+
+  public LogEventService(
+      QueryServiceClient queryServiceClient,
+      int requestTimeout,
+      AttributeMetadataProvider attributeMetadataProvider) {
+    this.queryServiceClient = queryServiceClient;
+    this.requestTimeout = requestTimeout;
+    this.attributeMetadataProvider = attributeMetadataProvider;
+    initMetrics();
+  }
+
+  private void initMetrics() {
+    queryExecutionTimer =
+        PlatformMetricsRegistry.registerTimer(
+            "hypertrace.log.event.query.execution", ImmutableMap.of());
+  }
+
+  public LogEventResponse getLogEventsByFilter(RequestContext context, LogEventRequest request) {
+    Instant start = Instant.now();
+    try {
+      Map<String, AttributeMetadata> attributeMap =
+          attributeMetadataProvider.getAttributesMetadata(context, "LOG_EVENT");
+      LogEventResponse.Builder logEventResponseBuilder = LogEventResponse.newBuilder();
+
+      Collection<LogEvent> filteredLogEvents = filterLogEvents(context, request, attributeMap);
+
+      logEventResponseBuilder.addAllLogEvents(filteredLogEvents);
+
+      LogEventResponse response = logEventResponseBuilder.build();
+      LOG.debug("Log Event Service Response: {}", response);
+      return response;
+    } finally {
+      queryExecutionTimer.record(
+          Duration.between(start, Instant.now()).toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  @VisibleForTesting
+  List<LogEvent> filterLogEvents(
```

Comment:
This method name sounds like it's just filtering a passed list of events, but it's actually the entire query execution - filtering, limiting, ordering etc.

Also, visible for testing is an antipattern - if we think this is worth directly testing, it should be in its own class - else, test through the public api.

Reply:
Updated